### PR TITLE
[cacheable-request] Include type alias for cache adapter

### PR DESCRIPTION
**Pull request by paulmelnikow**
This allows easily declaring a parameter that will be passed into the
constructor.

Ref https://github.com/sindresorhus/got/pull/760#discussion_r268277506

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ ] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Did not add a test because existing tests seem to cover this behavior.

Lint is failing, but I don't understand why:

```
> definitely-typed@0.0.3 lint /Users/pnm/code/DefinitelyTyped
> dtslint types "cacheable-request"

/Users/pnm/code/DefinitelyTyped/node_modules/.bin/dtslint: line 1: use strict: command not found
/Users/pnm/code/DefinitelyTyped/node_modules/.bin/dtslint: line 2: syntax error near unexpected token `('
/Users/pnm/code/DefinitelyTyped/node_modules/.bin/dtslint: line 2: `var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {'
```

Tests will not run on my machine either:

```
index.d.ts:17:1 - error TS2309: An export assignment cannot be used in a module with other exported elements.

17 export = CacheableRequest;
   ~~~~~~~~~~~~~~~~~~~~~~~~~~
```

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Comment by paulmelnikow** [comment]
I fixed the style error. The other error looks like a real error, though it's not related to my changes and I don't know the correct way to fix it. I'd appreciate any advice on that!

```
ERROR: 17:1   expect     TypeScript@next compile error: 
An export assignment cannot be used in a module with other exported elements.
```

File: types/cacheable-request/index.d.ts [hunk 1]
```diff
@@ -19,8 +20,10 @@ declare const CacheableRequest: CacheableRequest;
 
 type RequestFn = typeof request;
 
+export type StorageAdapter = Store<any>;
```
Comment (BendingBender) [inline types/cacheable-request/index.d.ts]: Better:
```ts
export {Store} from 'keyv';
```

File: types/cacheable-request/index.d.ts [hunk 2]
```diff
@@ -19,8 +20,10 @@ declare const CacheableRequest: CacheableRequest;
...
+
 interface CacheableRequest {
-    new (requestFn: RequestFn, storageAdapter?: string | Store<any>): (
+    new (requestFn: RequestFn, storageAdapter?: string | StorageAdapter): (
```
Comment (BendingBender) [inline types/cacheable-request/index.d.ts]: Revert this change.

**Review by BendingBender — CHANGES_REQUESTED** [review]

File: types/cacheable-request/index.d.ts (hunk 1 above)
Comment (paulmelnikow) [reply types/cacheable-request/index.d.ts]: Ah, thanks for explaining that!

I moved it into the namespace, though I left the templated type in there to keep this streamlined, since as far as I can tell, it's not changeable by the caller.

File: types/cacheable-request/index.d.ts [hunk 3]
```diff
@@ -33,6 +31,8 @@ interface CacheableRequest {
 }
 
 declare namespace CacheableRequest {
+    type StorageAdapter = Store<any>;
```
Comment (BendingBender) [inline types/cacheable-request/index.d.ts]: Please don't hard-code the `any` type parameter in the `StorageAdapter` type to allow users of this type to use this parameter to correctly align `get()` and `set()` signatures of the `StorageAdapter` implementation:

```suggestion
    type StorageAdapter<TValue> = Store<TValue>;
```

Don't forget to set this parameter to `any` in the CacheableRequest constructor argument.

**Review by BendingBender — CHANGES_REQUESTED** [review]

File: types/cacheable-request/index.d.ts (hunk 3 above)
Comment (paulmelnikow) [reply types/cacheable-request/index.d.ts]: Isn't the value type determined by the library? It's not as if the caller can meaningfully declare any type they choose; for their type to be helpful it needs to conform to the library's usage:

```js
{
  cachePolicy: response.cachePolicy.toObject(),
  url: response.url,
  statusCode: response.fromCache ? revalidate.statusCode : response.statusCode,
  body
}
```

If we don't use `Store<any>` it seems like it would be better to add an explicit type for the cache value.

Comment (BendingBender) [reply types/cacheable-request/index.d.ts]: If you would add an explicit type for the cache value, that would be great! Otherwise, stick with what you've done so far.

**Review by BendingBender — APPROVED** [review]

File: types/cacheable-request/index.d.ts (hunk 3 above)
Comment (paulmelnikow) [reply types/cacheable-request/index.d.ts]: Thanks, I'm happy to address that in a follow-on.
